flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java

```diff
@@ -503,23 +644,30 @@ public void close() throws Exception {
 	public void initializeState(FunctionInitializationContext context) throws Exception {
 
 		OperatorStateStore stateStore = context.getOperatorStateStore();
-		offsetsStateForCheckpoint = stateStore.getSerializableListState(DefaultOperatorStateBackend.DEFAULT_OPERATOR_STATE_NAME);
 
-		if (context.isRestored()) {
-			if (restoredState == null) {
-				restoredState = new HashMap<>();
-				for (Tuple2<KafkaTopicPartition, Long> kafkaOffset : offsetsStateForCheckpoint.get()) {
-					restoredState.put(kafkaOffset.f0, kafkaOffset.f1);
-				}
+		ListState<Tuple2<KafkaTopicPartition, Long>> oldRoundRobinListState =
+			stateStore.getSerializableListState(DefaultOperatorStateBackend.DEFAULT_OPERATOR_STATE_NAME);
 
-				LOG.info("Setting restore state in the FlinkKafkaConsumer.");
-				if (LOG.isDebugEnabled()) {
-					LOG.debug("Using the following offsets: {}", restoredState);
-				}
+		this.unionOffsetStates = stateStore.getUnionListState(new ListStateDescriptor<>(
+				OFFSETS_STATE_NAME,
+				TypeInformation.of(new TypeHint<Tuple2<KafkaTopicPartition, Long>>() {})));
+
+		if (context.isRestored() && !restoredFromOldState) {
+			restoredState = new TreeMap<>(new KafkaTopicPartition.Comparator());
+
+			// migrate from 1.2 state, if there is any
+			for (Tuple2<KafkaTopicPartition, Long> kafkaOffset : oldRoundRobinListState.get()) {
+				restoredFromOldState = true;
```

Comment:
You're right, this is definitely problematic. I think we'll have to block this feature until we have the required state information accessible before we can merge this then.